readium/streamer/src/main/java/org/readium/r2/streamer/parser/epub/MetadataAdapter.kt

```diff
@@ -0,0 +1,465 @@
+/*
+ * Copyright 2022 Readium Foundation. All rights reserved.
+ * Use of this source code is governed by the BSD-style license
+ * available in the top-level LICENSE file of the project.
+ */
+
+package org.readium.r2.streamer.parser.epub
+
+import org.readium.r2.shared.extensions.iso8601ToDate
+import org.readium.r2.shared.extensions.toMap
+import org.readium.r2.shared.publication.*
+import org.readium.r2.shared.publication.Collection
+import org.readium.r2.shared.publication.presentation.Presentation
+
+internal class MetadataAdapter(
+    private val epubVersion: Double,
+    private val fallbackTitle: String,
+    private val uniqueIdentifierId: String?,
+    private val readingProgression: ReadingProgression,
+    private val displayOptions: Map<String, String>
+) {
+    data class Result(
+        val links: List<Link>,
+        val metadata: Metadata,
+        val durationById: Map<String, Double?>,
+        val coverId: String?
+    )
+
+    fun adapt(items: List<MetadataItem>): Result {
+        val (globalItems: List<MetadataItem>, refiningItems: List<MetadataItem>) =
+            items.partition { it.refines == null }
+
+        @Suppress("Unchecked_cast")
+        val durationById = refiningItems
+            .groupBy(MetadataItem::refines)
+            .mapValues { DurationAdapter().adapt(it.value).first }
+            as Map<String, Double?>
+
+        var remainingItems: List<MetadataItem> = globalItems
+
+        val coverId: String? = remainingItems
+            .takeFirstWithProperty("cover")
+            .let { remainingItems = it.second; it.first }
+            ?.value
+
+        val duration: Double? = DurationAdapter()
+            .adapt(remainingItems)
+            .let { remainingItems = it.second; it.first }
```

Comment:
I don't find this pattern very readable. Maybe we can improve it?

We could modify in place a `MutableList` but if you prefer keeping the immutability:

There are a quite a lot of adapters following this template:

```kotlin
fun adapt(items: List<MetadataItem>): Pair<XXX, List<MetadataItem>>
```

Let's say we add a `MetadataItemAdapter` function interface:

```kotlin
data class MetadataItemResult<T> {
    val result: T
    val remaining: List<MetadataItem>
}

fun interface MetadataItemAdapter<T> {
    fun adapt(items: List<MetadataItem>): MetadataItemResult<T>
}
```

Then we should be able to add an inline `extract` helper:

```kotlin
var remainingItems: List<MetadataItem> = globalItems

fun <T> extract(adapter: MetadataItemAdapter<T>): T =
    adapter.adapt(remainingItems)
        .let { remainingItems = it.remaining; it.result }

val duration: Double? = extract(DurationAdapter())
val languages: List<String> = extract(LanguageAdapter(readingProgression))
```

Reply:
Yeah, I was not very happy with this pattern. I couldn't find a better one though.
If I understand correctly your proposal, we would have to define `extract` in every place where a `remainingItems` list is created.

Mutability in a very small and specialized class might be reasonable. 

```kotlin
data class MetadataItemsHolder(var items: List<MetadataItem>) {
    fun <T> adapt(adapter: MetadataAdapter<T>): T {
        adapter.adapt(items)
              .let { items = it.remaining; it.result }
    }
}
```

Reply:
> If I understand correctly your proposal, we would have to define extract in every place where a remainingItems list is created.

That's right.

> Mutability in a very small and specialized class might be reasonable.

I agree, that looks good.